# Create simple agent

## What is an Agent?

Think of an **Agent** like a smart robot assistant. You give it:
- A **brain** → the AI model (e.g. Llama, GPT)
- **Hands** → tools it can use (e.g. check weather, search web)
- **Instructions** → a system prompt telling it what role to play

The agent decides on its own **which tool to use** and **when to use it** to answer your question.

### How it works step by step:
1. User asks: *"What is weather in Pune?"*
2. Agent thinks: *"I need to call `get_weather` with city='pune'"*
3. Tool runs and returns: *"Weather in pune is sunny"*
4. Agent replies back to user in natural language

> **Key things to notice in the code below:**
> - `@tool` turns a normal Python function into a tool the agent can use
> - Always write a docstring (`"""..."""`) — the AI reads it to understand what the tool does
> - `response['messages'][-1].content` gets the **last message** (the agent's final reply)

In [2]:
import os
from dotenv import  load_dotenv
load_dotenv()


True

✅ **Output:** `"The city of Pune is currently sunny."`

The agent called `get_weather("pune")` automatically and used the result to form a natural reply.

In [3]:
from langchain.agents import  create_agent
from langchain.tools import  tool

@tool
def get_weather(city:str) -> str:
    """Get the weather for city""" 
    return f"Weather in  {city} is sunny"

agent = create_agent("groq:llama-3.1-8b-instant", tools=[get_weather], name="test agent", system_prompt="You are helpful assistant")

response = agent.invoke({
    "messages": [{"role": "user", "content": "What is weather in pune" }]
})
response['messages'][-1].content

'Since it keeps repeating the same thing that is weather in pune is sunny.'

## get metadata


Every AI response uses **tokens** (like currency). You can check how many were used via `usage_metadata`:

In [6]:
from langchain.agents import  create_agent
from langchain.tools import  tool

@tool
def get_weather(city:str) -> str:
    """Get the weather for city""" 
    return f"Weather in  {city} is sunny"

agent = create_agent("groq:llama-3.1-8b-instant", tools=[get_weather], name="test agent", system_prompt="You are helpful assistant")

response = agent.invoke({
    "messages": [{"role": "user", "content": "What is weather in pune" }]
})
response['messages'][-1].usage_metadata

{'input_tokens': 344, 'output_tokens': 8, 'total_tokens': 352}

✅ **Output:** `{'input_tokens': 344, 'output_tokens': 8, 'total_tokens': 352}`

| Field | Meaning |
|-------|---------|
| `input_tokens` | Tokens in your question + tool descriptions |
| `output_tokens` | Tokens in the AI's reply |
| `total_tokens` | Total used (fewer = cheaper & faster) |